### ISSUE TITLE
Simple 3d object transforms (translate/rotate/scale)

### ISSUE THREAD

**Issue by straaljager** [BlueBrain/Brayns#322]
Request from Christian:

"Ability to apply a transformation to scene morphologies and meshes (rotation, translation, scaling)"

This would require:
- object picking
- highlight the active object (e.g. by showing a bounding box in WebGL or highlighting the name in a list of scene objects)
- a WebGL 3d transform manipulator widget (like https://threejs.org/examples/#misc_controls_transform)

**Comment by hernando** [comment]
The request from Christian doesn't specify it has to be done using a UI, it can be done using code and in fact I think we would prefer it that way.

**Comment by straaljager** [comment]
True, WebGL overlay may not be needed

**Comment by rolandjitsu** [comment]
@straaljager I'd strongly disagree to use a different renderer (WebGL) in parallel with the existing renderer as we will get into synchronisation issues due to latency.

And the complexity will also increase because both the client and the backend would need to know about the model/data if we use WebGL, whereas leaving all the work to the backend (Brayns) it's much easier.

**Comment by hernando** [comment]
The point of having a WebGL rendering reusing the depth buffer is that the user can have faster feedback (lower latency) on where the objects are going to be placed. If the rendering can't keep up with the interaction, I wouldn't find disturbing the latency between the widget and the real rendering. In fact, I would find a lot more annoying not having visual feedback about the widget location in real-time. That said, I've tried to search if anybody has done something like this before, but I haven't been able to find anything.  
Other solution could be to simplify the rendering while interacting with the the scene, but this also has it's own challenges.

In any case, I think this is very subjective and without having an implementation that users can't experiment with we cannot claim whether this or that is generally preferred.

**Comment by rolandjitsu** [comment]
@hernando it's true that as a user you want feedback as soon as you manipulate, but having feedback that is out of sync or incoherent is not preferable and neither helpful.

It also means that if you want to render the same model/scene on the client side you need the data in order to render it. But this comes with at least two drawbacks:

- you have 2 rendering engines that you need to maintain
- data will need to be parsed on the client side and for large models that might not be possible due to [memory limitations](https://stackoverflow.com/a/32671934/1092007) or it might be too slow

It might work for very simple scenes, in which case you don't need a backend and pretty much everything can be done on the client side, but it will not work for complex scenes/large data sets.

It is indeed subjective, but I can assure you that going this way will only increase the complexity of the applications. It's much better to have the rendering and processing of data running on the backend, and that applies for pretty much most web based applications.

**Comment by straaljager** [comment]
Downloading to the full scene to client is not needed, just a proxy or even the bounding boxes of objects, like in 3ds Max should suffice. 

In my previous job we built a web application which did exactly that: rendering the full res model on a remote cluster with photorealistic ray tracing and streaming the images to the browser while having an extremely simplified version of the scene in WebGL that was used for interaction. The WebGL window could be rendered side by side or as an overlay. It could show which objection the raytraced viewport was active by displaying its bounding box or rectangle. 

Lagoa Render does it too for example: https://www.youtube.com/watch?v=MPJTbD9DgtU

**Comment by rolandjitsu** [comment]
@straaljager sure, but this is for very simply scenes. And there is already a WebGL impl. that handles scene management for simple scenes which is done by the Neurorobotics team.

I think that our case is a bit different, we don't build cars and/or other small scale objects, we will presumably work with millions of neurons which will fill up the browser memory fast (note that WebGL runs on the GPU so memory is even more limited or not even supported by most notebooks). And I'm not sure how you can simplify the scenes for hundreds of thousands of points and connections in a way that it would make sense to the user and also how valuable that is anyway.

I also don't see the value of using WebGL at all since the current latency is low and user feedback is close to immediate (especially for small scenes), event for relatively large data sets.

And as @tribal-tec mentioned, I don't think that for building circuits the user will feel comfortable manipulating 10k+ (or whatever number exceeding 100) neurons in the UI. Preferably there should be a programmatic API  which would allow the user to do this easily.

**Comment by straaljager** [comment]
Perhaps WebGL is not needed if the latency is very low anyway. But how will you visualise helpers like slicing planes to show a hippocampal slice for example? It could perhaps be done by adding a dynamic plane to the ray traced scene. 
We also need a way to draw selection regions, which should be a 2D overlay on the image on the client side.

**Comment by hernando** [comment]
@rolandjitsu As Sam already pointed out, the WebGL rendering would be limited to bounding boxes, planes or very other simple widgets . I wouldn't push any geometrical data to the client side for the reasons you have already mentioned. My impression is that the cases in which you want to place cells using 3D interaction reduce to scenes with at most hundreds of cells or using more advanced manipulations yet to be defined, but this is all speculation. As I already mentioned in my very first comment (and we agree in this point), the first thing to implement is an API, which is what our users have been using at the moment in RTNeuron.
2D Overlays shouldn't be discarded because they can be used to do rectangular selections, show labels and plots, but this is a separate discussion anyway.

**Comment by rolandjitsu** [comment]
Closing in favour of https://github.com/BlueBrain/Brayns/issues/363.
